rfcs/0112-demote-darwin-support-tier.md

```diff
@@ -0,0 +1,91 @@
+---
+feature: demote_darwin_support_tier
+start-date: 2021-10-27
+author: piegames
+co-authors: many
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Move x86_64-darwin from a Tier 2 platform to Tier 3, as described in [RFC 0046](https://github.com/NixOS/rfcs/blob/master/rfcs/0046-platform-support-tiers.md).
+
+# Motivation
+[motivation]: #motivation
+
+Darwin is constantly adding additional maintainer burden. Especially, it does not live up to the requirements:
+
+- "A lot of packages built by Hydra, full ofBorg support."
+    - There is only one ofBorg builder. It can take several hours or days for an ofBorg build to finish for Darwin, if it finishes at all.
+    - Hydra builders a greatly understaffed.
```

Comment:
* We have lacked build capacity for darwin in the past in terms of number of builders and this may very well be a problem in the future again. For darwin we have to use physical machines ordered by the foundation and set up manually, due to legal and organizational issues. This approach scales much worse than what we use for linux currently and would be much harder to solve since it requires physical access.
* There seem to be specific issues with either darwin or the darwin builders we have that is not down to a lack of people in the infra team. The darwin builders tend to get stuck for still unknown (?) reasons which requires reboots or restarting the queue runners. Such issues rarely happen with `aarch64-linux` and `x86_64-linux`. I don't find it very compelling to solve such an issue by throwing people at a problem that shouldn't exist in the first place.

Reply:
Re your second bullet. You say the problem is not due to a lack of people on the infra team but how do you expect the problem to be solved if no one can look into it? AFAIK the only people who can trigger the stop-gap "solution" of restarting builders and queue runners are Eelco and Graham. It seems like they either don't have time to do so or have another reason for not doing so (does it mess with progress of Linux evaluations?).

These same people are also the only ones with enough access to look into the problem, figure out these unknown reasons. If they don't have the time to restart the queue runner how can we expect them to have enough time to look into this? Many people have asked publicly whether they can help, I've also asked privately. Responses haven't been forthcoming. I believe the main problem here is there's no established way to deal with the trust required, maybe limited or temporary access could help with this?

We are trying to improve the Darwin situation to make it less of a burden to core maintainers and not having reliable CI makes this way harder on all parties involved.

Reply:
> For darwin we have to use physical machines ordered by the foundation and set up manually, due to legal and organizational issues.

Is there more context available on this? AWS and MacStadium provide hosted machines for example, and they have enough support from Apple that they had access to M1 machines before they were generally available, so I can't imagine they're doing something Apple doesn't approve of. Maybe the situation changed since darwin ofborg was initially set up?

Reply:
AWS macs are actually physical machines and are really very expensive.

Reply:
AFAIU the main problem is these machines can only be set up rather imperatively. So they're hard to spin up/down. That's why the Darwin builders are VMs running on Nixos running on Mac Minis (at least as far as x86_64 is concerned, I expect the M1 Minis aren't virtualized yet?).

Reply:
Does the fact that the `x86_64-darwin` builders are running in VMs, rather than macOS on bare metal, have anything to do with why they get stuck? Assuming the M1 builders aren't virtualized, are they hanging with the same frequency as the Intel builders?

Reply:
I don't think it's the builders getting stuck. They usually become idle. It's the queue runner that stops scheduling jobs for some reason.

Reply:
And this only affects the macOS builders, and not the NixOS builders, is that right?

Reply:
That's my understanding. Though this is all gleaned from messages here and in various Matrix rooms. Maybe @grahamc or @vcunat or others have more accurate information?

Reply:
The most common issue IIRC is that the builders get stuck on a single build step for days, the log often just showing “Sending Inputs”. One fix has been rebooting these machines.